pkg/engine/internal/planner/logical/planner.go

```diff
@@ -186,101 +188,161 @@ func buildPlanForLogQuery(
-			//case syntax.OpTypeMax:
-			//	vecAggType = types.VectorAggregationTypeMax
-			//case syntax.OpTypeMin:
-			//	vecAggType = types.VectorAggregationTypeMin
-			default:
-				err = errUnimplemented
-				return false
-			}
+	return rangeAggregation, nil
+}
 
-			groupBy = make([]ColumnRef, 0, len(e.Grouping.Groups))
-			for _, group := range e.Grouping.Groups {
-				groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
-			}
+func walkVectorAggregation(e *syntax.VectorAggregationExpr, params logql.Params) (Value, error) {
+	// `without()` grouping is not supported.
+	if e.Grouping != nil && e.Grouping.Without {
+		return nil, errUnimplemented
+	}
 
-			return true
-		default:
-			err = errUnimplemented
-			return false // do not traverse children
-		}
-	})
+	left, err := walk(e.Left, params)
 	if err != nil {
 		return nil, err
 	}
 
-	if rangeAggType == types.RangeAggregationTypeInvalid || vecAggType == types.VectorAggregationTypeInvalid {
+	vecAggType := convertVectorAggregationType(e.Operation)
+	if vecAggType == types.VectorAggregationTypeInvalid {
 		return nil, errUnimplemented
 	}
 
-	logSelectorExpr, err := e.Selector()
+	groupBy := make([]ColumnRef, 0, len(e.Grouping.Groups))
+	for _, group := range e.Grouping.Groups {
+		groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
+	}
+
+	return &VectorAggregation{
+		Table:     left,
+		GroupBy:   groupBy,
+		Operation: vecAggType,
+	}, nil
+}
+
+func hasNonMathExpressionChild(n Value) bool {
+	if _, ok := n.(*VectorAggregation); ok {
+		return true
+	}
+
+	if _, ok := n.(*RangeAggregation); ok {
+		return true
+	}
+
+	if b, ok := n.(*BinOp); ok {
+		return hasNonMathExpressionChild(b.Left) || hasNonMathExpressionChild(b.Right)
+	}
+
+	return false
+}
+
+func walkBinOp(e *syntax.BinOpExpr, params logql.Params) (Value, error) {
+	left, err := walk(e.SampleExpr, params)
 	if err != nil {
 		return nil, err
 	}
-
-	builder, err := buildPlanForLogQuery(logSelectorExpr, params, true, rangeInterval)
+	right, err := walk(e.RHS, params)
 	if err != nil {
 		return nil, err
 	}
 
-	builder = builder.RangeAggregation(
-		nil, rangeAggType, params.Start(), params.End(), params.Step(), rangeInterval,
-	).VectorAggregation(groupBy, vecAggType)
+	op := convertBinaryArithmeticOp(e.Op)
+	if op == types.BinaryOpInvalid {
+		return nil, errUnimplemented
+	}
+
+	// this is to check that there is only one non-literal input on either side, otherwise it is not implemented yet.
+	// TODO remove when all MathExpression pipeline can read multiple inputs
+	if hasNonMathExpressionChild(left) && hasNonMathExpressionChild(right) {
+		return nil, errUnimplemented
+	}
+
+	return &BinOp{
+		Left:  left,
+		Right: right,
+		Op:    op,
+	}, nil
+}
+
+func walkLiteral(e *syntax.LiteralExpr, _ logql.Params) (Value, error) {
+	return &Literal{
+		NewLiteral(e.Val),
+	}, nil
+}
+
+func walk(e syntax.Expr, params logql.Params) (Value, error) {
```

Comment:
nit: `e`'s type could be narrowed down to `syntax.SampleExpr`, which would help us make sure we have full coverage here. 

Also, since `walk` is only used for sample queries, `walkSampleExpr` might be a more fitting name.